### PR TITLE
Editorial: Correct notes in Create{Method,NonEnumerableData}Property 

### PR DESCRIPTION
**Pull request by linusg**
Details in the commit messages.

### CODE REVIEW

**Comment by arhadthedev** [comment]
Should the order of the AOs be returned back (`CreateDataProperty` then `CreateNonEnumerableDataProperty`)? The original looks alphabetically sorted.

**Comment by linusg** [comment]
The current order is

```
7.3.5 CreateDataProperty ( O, P, V )
7.3.6 CreateMethodProperty ( O, P, V )
7.3.7 CreateDataPropertyOrThrow ( O, P, V )
7.3.8 CreateNonEnumerableDataPropertyOrThrow ( O, P, V )
7.3.9 DefinePropertyOrThrow ( O, P, desc )
7.3.10 DeletePropertyOrThrow ( O, P )
```

...i.e. not alphabetical. My intention was to keep "...OrThrow" last.

**Comment by linusg** [comment]
The current order on `main` is

```
7.3.5 CreateDataProperty ( O, P, V )
7.3.6 CreateMethodProperty ( O, P, V )
7.3.7 CreateDataPropertyOrThrow ( O, P, V )
7.3.8 CreateNonEnumerableDataPropertyOrThrow ( O, P, V )
7.3.9 DefinePropertyOrThrow ( O, P, desc )
7.3.10 DeletePropertyOrThrow ( O, P )
```

...i.e. not alphabetical. My intention was to keep "...OrThrow" last.

File: spec.html [hunk 1]
```diff
@@ -6348,52 +6348,58 @@ <h1>
...
       <dl class="header">
         <dt>description</dt>
-        <dd>It is used to create a new own property of an object. It throws a *TypeError* exception if the requested property update cannot be performed.</dd>
+        <dd>It is used to create a new non-enumerable own property of an ordinary object.</dd>
       </dl>
       <emu-alg>
-        1. Let _success_ be ? CreateDataProperty(_O_, _P_, _V_).
-        1. If _success_ is *false*, throw a *TypeError* exception.
-        1. Return _success_.
+        1. Assert: _O_ is an ordinary, extensible object with no non-configurable properties.
```
Comment (ljharb) [inline spec.html:6371]: How can this assertion be made? This spec isn’t the only thing that calls these AOs.

**Comment by arhadthedev** [comment]
@linusg Now I see, thank you. I didn't look through the whole section initially.

File: spec.html [hunk 2]
```diff
@@ -6367,12 +6370,15 @@ <h1>
       <emu-alg>
         1. Assert: _O_ is an ordinary, extensible object with no non-configurable properties.
         1. Let _newDesc_ be the PropertyDescriptor { [[Value]]: _V_, [[Writable]]: *true*, [[Enumerable]]: *false*, [[Configurable]]: *true* }.
-        1. Perform ! DefinePropertyOrThrow(_O_, _P_, _newDesc_).
+        1. Perform ! _O_.[[DefineOwnProperty]](_P_, _newDesc_).
```
Comment (ljharb) [inline spec.html:6373]: I find this reduces clarity; calling an AO a better than calling an internal method.

**Comment by bakkot** [comment]
Thanks for the PR. I don't think the renaming is an improvement; the "or throw" is intentional, for consistency with `CreateDataPropertyOrThrow`. We do need to clean these up (cf https://github.com/tc39/ecma262/pull/2528) but I don't think we want to make that particular change.

The refactoring to call `[[DefineOwnProperty]]` directly also does not seem like an improvement improvement, since it now ignores the return value instead of checking it. `! DefinePropertyOrThrow()` is the pattern we use throughout the specification when defining a non-accessor property in a way where we know that will succeed.

Second commit seems fine, if you want to reduce this PR to just that.

File: spec.html (hunk 1 above)
Comment (linusg) [reply spec.html:6371]: I didn't make that change, it only shows as new from the re-ordering. This is already in the current spec.

**Comment by linusg** [comment]
> the "or throw" is intentional, for consistency with `CreateDataPropertyOrThrow`.

`CreateDataPropertyOrThrow` actually *can* throw. `CreateNonEnumerableDataPropertyOrThrow` *can't* as written, so the name is super misleading. It's functionally equivalent to `CreateMethodProperty`, which isn't named `CreateMethodPropertyOrThrow` either. I don't think consistency is applicable there?

> The refactoring to call [[DefineOwnProperty]] directly also does not seem like an improvement, since it now ignores the return value instead of checking it.

That however was purely for consistency, it's the same way `CreateMethodProperty` works. Would you accept a change to update that to call the `DefinePropertyOrThrow` AO instead of directly the internal method?

> ! DefinePropertyOrThrow() is the pattern we use throughout the specification when defining a non-accessor property in a way where we know that will succeed.

I know, and it still does that (`! O.[[DefineOwnProperty]](...)`). The only difference is less indirection.

> Second commit seems fine, if you want to reduce this PR to just that.

Happy to do that, but just want to confirm that there's no misunderstanding here. Keeping the current name seems like a rather unfortunate decision, and the PR you mentioned doesn't seem to cover that either.

File: spec.html (hunk 2 above)
Comment (linusg) [reply spec.html:6373]: Thanks, I'll keep that in mind :+1:

**Comment by bakkot** [comment]
>  I don't think consistency is applicable there?

I agree there's already inconsistency, but I'm not a fan of this particular fix for the inconsistency.

> Would you accept a change to update that to call the `DefinePropertyOrThrow` AO instead of directly the internal method?

Sure, that sounds good.

> I know, and it still does that (`! O.[[DefineOwnProperty]](...)`). The only difference is less indirection.

No, `! DefinePropertyOrThrow(O, ...)` is meaningfully different than `! O.[[DefineOwnProperty]](...)`: the former contains an assertion that creation of the property succeeds, whereas the latter does not. (`[[DefineOwnProperty]]` returns `*false*` on failure rather than throwing.)

> Keeping the current name seems like a rather unfortunate decision, and the PR you mentioned doesn't seem to cover that either.

The PR I mentioned isn't what we actually want to do, I just linked it to point to where (some) discussion has happened previously. Editors are agreed we need to rename and clean up these AOs but have not yet actually figured out what we want that to look like.

**Comment by linusg** [comment]
Thanks for the feedback all, updated!

**Review by bakkot — APPROVED** [review]

File: spec.html [hunk 3]
```diff
@@ -6344,11 +6344,11 @@ <h1>
       <emu-alg>
         1. Assert: _O_ is an ordinary, extensible object with no non-configurable properties.
         1. Let _newDesc_ be the PropertyDescriptor { [[Value]]: _V_, [[Writable]]: *true*, [[Enumerable]]: *false*, [[Configurable]]: *true* }.
-        1. Perform ! _O_.[[DefineOwnProperty]](_P_, _newDesc_).
+        1. Perform ! DefinePropertyOrThrow(_O_, _P_, _newDesc_).
         1. Return ~unused~.
       </emu-alg>
       <emu-note>
-        <p>This abstract operation creates a property whose attributes are set to the same defaults used for built-in methods and methods defined using class declaration syntax. Normally, the property will not already exist. If it does exist and is not configurable or if _O_ is not extensible, [[DefineOwnProperty]] will return *false*.</p>
+        <p>This abstract operation creates a property whose attributes are set to the same defaults used for built-in methods and methods defined using class declaration syntax. Normally, the property will not already exist. If it does exist, [[DefineOwnProperty]] is guaranteed to complete normally and return *true* within this specification.</p>
```
Comment (michaelficarra) [inline spec.html:6351]:
```suggestion
        <p>This abstract operation creates a property whose attributes are set to the same defaults used for built-in methods and methods defined using class declaration syntax. Normally, the property will not already exist. If it does exist, DefinePropertyOrThrow is guaranteed to complete normally.</p>
```

File: spec.html [hunk 4]
```diff
@@ -6375,7 +6375,7 @@ <h1>
         1. Return ~unused~.
       </emu-alg>
       <emu-note>
-        <p>This abstract operation creates a property whose attributes are set to the same defaults used for properties created by the ECMAScript language assignment operator except it is not enumerable. Normally, the property will not already exist. If it does exist and is not configurable or if _O_ is not extensible, [[DefineOwnProperty]] will return *false* causing this operation to throw a *TypeError* exception.</p>
+        <p>This abstract operation creates a property whose attributes are set to the same defaults used for properties created by the ECMAScript language assignment operator except it is not enumerable. Normally, the property will not already exist. If it does exist, [[DefineOwnProperty]] is guaranteed to complete normally and return *true* within this specification.</p>
```
Comment (michaelficarra) [inline spec.html:6378]: Same here, please.

```suggestion
        <p>This abstract operation creates a property whose attributes are set to the same defaults used for properties created by the ECMAScript language assignment operator except it is not enumerable. Normally, the property will not already exist. If it does exist, DefinePropertyOrThrow is guaranteed to complete normally.</p>
```

Comment (linusg) [reply spec.html:6378]: oops, fixed!

**Review by michaelficarra — APPROVED** [review]

**Comment by linusg** [comment]
Who do I ping to click the merge button? :^)

**Review by syg — APPROVED** [review]

**Pull request merged**